synapse/storage/persist_events.py

```diff
@@ -854,7 +855,14 @@ async def _prune_extremities(
                 and event.depth < current_depth - 100
             ):
                 continue
-            if get_domain_from_id(event.sender) in new_senders:
+
+            # We can be less conservative about dropping extremities from the
+            # same domain, though we do want to wait a little bit (otherwise
+            # we'll immediately remove all extremities from a given server).
+            if (
+                get_domain_from_id(event.sender) in new_senders
+                and event.depth < current_depth - 20
```

Comment:
this now seems to be in conflict with 'the extremity must give the same state as the "current state"'. In the rooms where this is a problem, it's relatively uncommon to make it 20 events without a state change. My concern is therefore that this whole algorithm will never actually be useful.

Why 20 here?

Reply:
> this now seems to be in conflict with 'the extremity must give the same state as the "current state"'. In the rooms where this is a problem, it's relatively uncommon to make it 20 events without a state change. My concern is therefore that this whole algorithm will never actually be useful.

State changes are fine, what we're looking for is that the resolved state with all extremities is the same as the resolved state without the old extremities. This should be true for simple gaps (rather than forks in the graph for example). 

> Why 20 here?

It's a fairly arbitrary low but non-zero number. What I saw on jki.re over the weekend is that we'd get two extremities from matrix.org and then immediately prune one, even though a new event will come along and point to both, which feels a bit wrong. Maybe using depth here is sort of the wrong thing.

Reply:
I wonder if we should also add a check for it being an hour old, or something? I guess we're saying dropping events shouldn't be problematic for remote events, but I could imagine a case where we have a lot of events sent all at once and so we end up with lots of extremities that will take some time to be referenced

Reply:
at this point we should probably just deploy something and maybe iterate on it?

Reply:
Yup, agreed